CDCgov/prime-reportstream
# Add Hl7 field repettion separation (~) handling to replaceValueAwithB

This PR add the hl7 field repetition separator (~) handling to the replaceValueAwithB setting feature.

Test Steps:
1. ./gradlew clean package
2. download the attached test file (pdi-covid-19-nh.csv) to [PRIME ROOT]/junk directory
3. Run: 
    $  ./prime data --input junk/pdi-covid-19-nh.csv --input-schema primedatainput/pdi-covid-19 --output-dir junk --route

4.) Without replaceValueAwithB apply to OBX-17 
Verify the hl7 output file should contains the OBX-17 as following:
OBX-17 Observation Method:
 ├ OBX-17.1 (Identifier)                        BD Veritor System for Rapid Detection of SARS-CoV-2_Becton, Dickinson and Company (BD)_EUA
 ├ OBX-17.2 (Text)                              BD Veritor System for Rapid Detection of SARS-CoV-2
 ├ OBX-17.3 (Name of Coding System)             99ELR
 ├ OBX-17.4 (Alternate Identifier)
 ├ OBX-17.5 (Alternate Text)
 ├ OBX-17.6 (Name of Alternate Coding System)
 ├ OBX-17.7 (undefined)                         2.68
 ├ OBX-17.8 (undefined)
 └ OBX-17.9 (undefined)                         BD Veritor System for Rapid Detection of SARS-CoV-2_Becton, Dickinson and Company (BD)_EUA

5.) Now, apply OBX-17 replaceValueAwithB by adding the following line to  organization.yml: -name: nh-dphs.receivers: -name:elr.translation: replaceValueAwithB:  It should look like as following:

- name: nh-dphs
  description: New Hampshire Division of Public Health Services
  jurisdiction: STATE
  stateCode: NH
  receivers:
    - name: elr
      organizationName: nh-dphs
      topic: covid-19
      customerStatus: active
      jurisdictionalFilter:
        - orEquals(ordering_facility_state, NH, patient_state, NH)
      translation:
        type: HL7
        useBatchHeaders: true
        messageProfileId: PHLabReport-Batch^^2.16.840.1.113883.9.11^ISO
        receivingFacilityName: NH_DHHS
        receivingFacilityOID: 2.16.840.1.114222.4.1.3669
        suppressHl7Fields: ORC-23-1,ORC-14-1,OBR-17-1,OBX-18-1, OBX-18-2, OBX-18-3, PID-13-1
        replaceValueAwithB:
	  MSH-3: [ "*": "CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO" ]
          MSH-4: [ "*": "CDC PRIME^11D2030855^CLIA" ]
          MSH-5: [ "*": "NH_ELR^2.16.840.1.114222.4.3.2.2.3.600.4^ISO" ]
          MSH-6: [ "*": "NH_DHHS^2.16.840.1.114222.4.1.3669^ISO" ]
          MSH-21: [ "*": "PHLabReport-Batch^^2.16.840.1.113883.9.11^ISO" ]
          OBX-17: [ "*": "DEVICE ID 1^^99ELR~DEVICE ID 2^^99ELR" ]

6.) Run
    $  ./prime data --input junk/pdi-covid-19-nh.csv --input-schema primedatainput/pdi-covid-19 --output-dir junk --route

7.) Verify the hl7 output file should contains the OBX-17 as following:
OBX-17[1] Observation Method:                   
 ├ OBX-17[1].1 (Identifier)                     DEVICE ID 1
 ├ OBX-17[1].2 (Text)                           
 ├ OBX-17[1].3 (Name of Coding System)          99ELR
 ├ OBX-17[1].4 (Alternate Identifier)           
 ├ OBX-17[1].5 (Alternate Text)                 
 ├ OBX-17[1].6 (Name of Alternate Coding System) 
 ├ OBX-17[1].7 (undefined)                      2.68
 ├ OBX-17[1].8 (undefined)                      
 └ OBX-17[1].9 (undefined)                      BD Veritor System for Rapid Detection of SARS-CoV-2_Becton, Dickinson and Company (BD)_EUA
OBX-17[2] Observation Method:                   
 ├ OBX-17[2].1 (Identifier)                     DEVICE ID 2
 ├ OBX-17[2].2 (Text)                           
 └ OBX-17[2].3 (Name of Coding System)          99ELR

Test File: 
[pdi-covid-19-nh.csv](https://github.com/CDCgov/prime-reportstream/files/8985069/pdi-covid-19-nh.csv)

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue
 https://app.zenhub.com/workspaces/onboarding--operations-6166edbd409257001e09f1b2/issues/cdcgov/prime-reportstream/5734

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
